rfcs/20190115-dense-attention.md

```diff
@@ -0,0 +1,506 @@
+    Returns:
+      Attention outputs of shape `[batch_size, Tq, dim]`.
+    """
+```
+
+The implementations for both Attention layers can be in the same file. They can
+reuse a private method with the following signature:
+
+```python
+def _apply_attention_scores(scores, value, value_mask=None):
+  """Applies attention scores to the given value tensor.
+
+  Args:
+    scores: Scores tensor of shape `[batch_size, Tq, Tv]`.
+    value: Value tensor of shape `[batch_size, Tv, dim]`.
+    value_mask: A boolean mask `Tensor` of shape `[batch_size, Tv]`.
+      If given, will apply the mask such that values at positions where
+      `mask==False` do not contribute to the result.
+
+  Returns:
+    Tensor of shape `[batch_size, Tq, dim]`.
+  """
+```
+
+Implementations of other Attention mechanisms can reuse this method, as well.
+So, that method can be made public. An alternative using inheritance is
+discussed in the "Base Attention Class" section. The
+rest of the code is specific to each Attention mechanism.
+
+Although not the primary focus of this proposal, the Attention layers work with
+RNN networks, such as
+[tf.keras.layers.LSTM](https://www.tensorflow.org/api_docs/python/tf/keras/layers/LSTM):
+When creating the LSTM, users need to set `return_sequences=True`, and the rest
+works the same way as CNN. It is unclear whether this method suffices to create
+the most common RNN+Attention models.
+
+We will first work on the implementation for Tensorflow.
+
+### Self-Attention
+
+The Self-Attention variant can be implemented by passing the same tensor to both
+`query` and `value`.
+
+### Multi-Head Attention
+
+This is an Attention variant proposed in
+["Attention is all you need"](https://arxiv.org/abs/1706.03762). This variant is
+not covered in our proposal. But could be implemented as an additional feature,
+e.g. by adding a `num_heads` argument that defaults to 1. The implementation
+will split the `query` and `value` tensors into `num_heads` tensors, calculate
+attention for each pair, then stack the results. This transformation can be
+implemented as a private method that is reused by all attention layers.
+
+### Transformer
+
+Transformer is a DNN+Attention network proposed in
+["Attention is all you need"](https://arxiv.org/abs/1706.03762). There is an
+implementation of it under
+https://github.com/tensorflow/models/tree/master/official/transformer, which
+uses a custom Attention implementation. Our proposal will simplify the
+Transformer network constructions, because users can reuse the Attention layers,
+rather than writing custom ones.
+
+### Position Representations
+
+DNN+Attention networks do not model relative or absolute position information in
+their structure. Instead, position information is modeled as an additional term
+in the model output. The proposed techniques can be implemented as an additional
+feature in the Attention API.
+
+*   https://arxiv.org/abs/1803.02155 describes how relative position
```

Comment:
It'd help to clarify here whether the position-dependent attention implementations compose with this layer or would require new layers.

Reply:
Good point. For relative position representation, we need to modify the attention layer. Absolute position can be implemented as a separate layer that composes with this layer. Let me update the text.